File: iii_drone_configuration/configurator.py
###############################################################################
# Imports
###############################################################################

from __future__ import annotations

from threading import Lock
from typing import Callable, Dict, List, Optional

import rclpy
from rclpy import lifecycle
from rclpy.qos import QoSProfile
from rcl_interfaces.msg import ParameterEvent, SetParametersResult
from std_msgs.msg import String

from .configuration import Configuration, ConfigurationEntry
from .schema_utils import resolve_schema_file


###############################################################################
# Class
###############################################################################


class Configurator:
    @staticmethod
    def _parameter_type_value(parameter_type: rclpy.parameter.Parameter.Type | int) -> int:
        return parameter_type.value if hasattr(parameter_type, "value") else int(parameter_type)

    def __init__(
        self,
        node: lifecycle.Node | rclpy.node.Node,
        after_parameter_change_callback: Optional[Callable[[rclpy.parameter.Parameter], None]] = None,
        qos: QoSProfile = QoSProfile(depth=10),
    ):
        self.node = node
        self.after_parameter_change_callback = after_parameter_change_callback
        self.qos = qos
        self.configurations: List[Configuration] = []
        self.parameters_mutex = Lock()
        self._managed_parameter_names: List[str] = []
        self.is_cleaned_up = False
        self._managed_node_announced = False

        self._declare_support_parameter_if_missing("parameters_path_postfix", "parameters/")
        self._declare_support_parameter_if_missing("default_parameter_file", "parameter_manifest.yaml")
        self._declare_support_parameter_if_missing("sim_parameter_file", "parameter_manifest.yaml")

        self.schema_file_path = resolve_schema_file(
            parameters_path_postfix=str(self.node.get_parameter("parameters_path_postfix").value),
            default_parameter_file=str(self.node.get_parameter("default_parameter_file").value),
            sim_parameter_file=str(self.node.get_parameter("sim_parameter_file").value),
        )
        try:
            from ._native import NativeConfiguratorCore
        except ImportError as exc:
            raise ImportError(
                "iii_drone_configuration native bindings are not available. "
                "Build the iii_drone_configuration package before using the Python Configurator."
            ) from exc
        self._native_core = NativeConfiguratorCore(str(self.schema_file_path))

        self.parameter_events_subscriber = self.node.create_subscription(
            ParameterEvent,
            "/parameter_events",
            self.parameter_event_callback,
            self.qos,
        )
        self._on_set_callback_handle = self.node.add_on_set_parameters_callback(self.on_set_parameters_callback)
        self._announcement_publisher = self.node.create_publisher(
            String,
            "/configuration/configuration_server/managed_node_available",
            10,
        )

    def _declare_support_parameter_if_missing(self, name: str, default_value: str) -> None:
        if not self.node.has_parameter(name):
            self.node.declare_parameter(name, default_value)

    def _announce_managed_node(self) -> None:
        if not self._managed_parameter_names:
            return
        msg = String()
        msg.data = self.node.get_fully_qualified_name()
        self._announcement_publisher.publish(msg)
        self._managed_node_announced = True

    def _current_managed_parameter_values(self) -> Dict[str, object]:
        values: Dict[str, object] = {}
        for parameter_name in self._managed_parameter_names:
            values[parameter_name] = self.node.get_parameter(parameter_name).value
        return values

    def _candidate_parameter_map(
        self,
        candidate_values: Dict[str, object],
        candidate_types: Dict[str, int],
    ) -> Dict[str, Dict[str, object]]:
        parameter_map = {}
        for name in self._native_core.schema_parameter_names():
            entry = self._native_core.get_schema_entry(name)
            parameter_map[name] = {
                "type": entry["parameter_type"],
                "value": entry["default_value"],
            }

        for name, value in candidate_values.items():
            parameter_map[name] = {
                "type": candidate_types[name],
                "value": value,
            }

        return parameter_map

    def cleanup(self):
        if self.is_cleaned_up:
            return

        if self.parameter_events_subscriber is not None:
            self.node.destroy_subscription(self.parameter_events_subscriber)
            self.parameter_events_subscriber = None

        if self._announcement_publisher is not None:
            self.node.destroy_publisher(self._announcement_publisher)
            self._announcement_publisher = None

        self.configurations.clear()
        self._managed_parameter_names.clear()
        self._native_core = None
        self.node = None
        self.is_cleaned_up = True

    def get_parameter(self, parameter_full_name: str) -> rclpy.parameter.Parameter:
        return self.get_parameters([parameter_full_name])[0]

    def get_parameters(self, parameter_full_names: List[str]) -> List[rclpy.parameter.Parameter]:
        with self.parameters_mutex:
            return [self.node.get_parameter(parameter_full_name) for parameter_full_name in parameter_full_names]

    def get_configuration(self, configuration_name: str) -> Configuration:
        for configuration in self.configurations:
            if configuration.name == configuration_name:
                return configuration
        raise RuntimeError(f"Configurator.get_configuration: Configuration '{configuration_name}' not found")

    def create_configuration(self, configuration_name: str, entries: List[ConfigurationEntry]) -> Configuration:
        configuration = Configuration(
            native_configuration=self._native_core.create_configuration(
                configuration_name,
                [(entry.full_name, self._parameter_type_value(entry.parameter_type)) for entry in entries],
            ),
            parameter_getter=lambda full_name: self.node.get_parameter(full_name),
        )
        self.configurations.append(configuration)
        return configuration

    @staticmethod
    def get_parameter_type_string(T: rclpy.parameter.Parameter.Type) -> str:
        from ._native import NativeConfiguratorCore
        return NativeConfiguratorCore.get_parameter_type_string(Configurator._parameter_type_value(T))

    @staticmethod
    def get_parameter_type_from_string(parameter_type_string: str) -> rclpy.parameter.Parameter.Type:
        from ._native import NativeConfiguratorCore
        return rclpy.parameter.Parameter.Type(NativeConfiguratorCore.get_parameter_type_from_string(parameter_type_string))

    def print_parameters(self):
        for param_name in self._managed_parameter_names:
            param = self.node.get_parameter(param_name)
            self.node.get_logger().info(f"Parameter: {param.name} = {param.value}")

    def print_configurations(self):
        for configuration in self.configurations:
            self.node.get_logger().info(f"Configuration: {configuration.name}")

    def validate(self):
        current_values = self._current_managed_parameter_values()
        if not current_values:
            return
        current_types = {
            name: self._parameter_type_value(self.node.get_parameter(name).type_)
            for name in self._managed_parameter_names
            if self.node.has_parameter(name)
        }
        self._native_core.validate_parameter_map(
            self._candidate_parameter_map(current_values, current_types),
            True,
        )
        if not self._managed_node_announced:
            self._announce_managed_node()

    def declare_parameter(
        self,
        parameter_full_name: str,
        parameter_type: rclpy.parameter.Parameter.Type,
    ) -> None:
        self.declare_parameters([parameter_full_name], [parameter_type])

    def declare_parameters(
        self,
        parameter_full_names: List[str],
        parameter_types: List[rclpy.parameter.Parameter.Type],
    ) -> None:
        if len(parameter_full_names) != len(parameter_types):
            raise RuntimeError("Configurator.declare_parameters: names/types size mismatch")

        with self.parameters_mutex:
            default_values = self._native_core.declare_parameters(
                parameter_full_names,
                [self._parameter_type_value(parameter_type) for parameter_type in parameter_types],
            )

            for name, default_value in zip(parameter_full_names, default_values):
                if not self.node.has_parameter(name):
                    self.node.declare_parameter(name, default_value)

                if name not in self._managed_parameter_names:
                    self._managed_parameter_names.append(name)

    def parameter_event_callback(self, parameter_event: ParameterEvent):
        if parameter_event.node != self.node.get_fully_qualified_name():
            return

        if self.after_parameter_change_callback is None:
            return

        for parameter_msg in parameter_event.changed_parameters:
            if parameter_msg.name in self._managed_parameter_names:
                self.after_parameter_change_callback(rclpy.parameter.Parameter.from_parameter_msg(parameter_msg))

    def on_set_parameters_callback(
        self,
        parameters: List[rclpy.parameter.Parameter],
    ) -> SetParametersResult:
        result = SetParametersResult(successful=True)
        candidate_values = self._current_managed_parameter_values()
        candidate_types = {
            name: self._parameter_type_value(self.node.get_parameter(name).type_)
            for name in self._managed_parameter_names
            if self.node.has_parameter(name)
        }

        for parameter in parameters:
            if parameter.name not in self._managed_parameter_names:
                continue

            try:
                candidate_values[parameter.name] = parameter.value
                candidate_types[parameter.name] = self._parameter_type_value(parameter.type_)
                self._native_core.validate_parameter_value(
                    parameter.name,
                    parameter.value,
                    self._parameter_type_value(parameter.type_),
                    self._candidate_parameter_map(candidate_values, candidate_types),
                    False,
                )
            except Exception as exc:
                result.successful = False
                result.reason = str(exc)
                return result

        try:
            self._native_core.validate_parameter_map(
                self._candidate_parameter_map(candidate_values, candidate_types),
                True,
            )
        except Exception as exc:
            result.successful = False
            result.reason = str(exc)
            return result

        return result

File: src/configurator.cpp
#include "iii_drone_configuration/configurator.hpp"

#include <ament_index_cpp/get_package_share_directory.hpp>

#include <rclcpp_lifecycle/lifecycle_node.hpp>

#include <rcl_interfaces/msg/parameter_descriptor.hpp>

#include <algorithm>
#include <cstdlib>
#include <filesystem>
#include <stdexcept>

using namespace iii_drone::configuration;

namespace {

std::string ExpandHome(const std::string & path)
{
    if (!path.empty() && path[0] == '~') {
        const char * home = std::getenv("HOME");
        if (home != nullptr) {
            return std::string(home) + path.substr(1);
        }
    }
    return path;
}

}  // namespace

template <typename nodeT>
Configurator<nodeT>::Configurator(
    nodeT *node,
    const std::string & node_name,
    std::function<void(const rclcpp::Parameter &)> after_parameter_change_callback
) : node_(node), after_parameter_change_callback_(after_parameter_change_callback)
{
    (void)node_name;

    schema_file_path_ = resolveSchemaFilePath();
    try {
        schema_validator_ = SchemaValidator::FromFile(schema_file_path_);
    } catch (const std::exception & ex) {
        throw std::runtime_error(
            "Configurator failed to load parameter schema '" + schema_file_path_ + "': " + ex.what()
        );
    }

    parameter_events_subscriber_ = node_->template create_subscription<rcl_interfaces::msg::ParameterEvent>(
        "/parameter_events",
        10,
        std::bind(&Configurator<nodeT>::parameterEventCallback, this, std::placeholders::_1)
    );

    managed_node_announcement_publisher_ = node_->template create_publisher<std_msgs::msg::String>(
        "/configuration/configuration_server/managed_node_available",
        10
    );

    on_set_parameters_callback_handle_ = node_->add_on_set_parameters_callback(
        std::bind(&Configurator<nodeT>::onSetParametersCallback, this, std::placeholders::_1)
    );

    if (!managed_parameter_names_.empty() && !managed_node_announced_) {
        announceManagedNode();
    }
}

template <typename nodeT>
Configurator<nodeT>::~Configurator()
{
    if (parameter_events_subscriber_) {
        parameter_events_subscriber_->clear_on_new_message_callback();
        parameter_events_subscriber_.reset();
    }

    if (on_set_parameters_callback_handle_) {
        node_->remove_on_set_parameters_callback(on_set_parameters_callback_handle_.get());
        on_set_parameters_callback_handle_.reset();
    }

    managed_parameter_names_.clear();
    configurations_.clear();
}

template <typename nodeT>
rclcpp::Parameter Configurator<nodeT>::GetParameter(const std::string & parameter_full_name) const
{
    return GetParameters({parameter_full_name}).front();
}

template <typename nodeT>
std::vector<rclcpp::Parameter> Configurator<nodeT>::GetParameters(const std::vector<std::string> & parameter_full_names) const
{
    std::shared_lock<std::shared_mutex> lock(parameters_mutex_);

    std::vector<rclcpp::Parameter> parameters;
    parameters.reserve(parameter_full_names.size());

    for (const auto & full_name : parameter_full_names) {
        rclcpp::Parameter parameter;
        if (!node_->get_parameter(full_name, parameter)) {
            throw std::runtime_error("Configurator::GetParameters(): Missing declared parameter " + full_name);
        }
        parameters.push_back(parameter);
    }

    return parameters;
}

template <typename nodeT>
Configuration::SharedPtr Configurator<nodeT>::GetConfiguration(const std::string & name) const
{
    for (const auto & configuration : configurations_) {
        if (configuration->name() == name) {
            return configuration;
        }
    }

    throw std::runtime_error("Configurator::GetConfiguration(): Configuration " + name + " does not exist.");
}

template <typename nodeT>
void Configurator<nodeT>::DeclareParameter(
    const std::string & parameter_full_name,
    rclcpp::ParameterType parameter_type
)
{
    DeclareParameters({parameter_full_name}, {parameter_type});
}

template <typename nodeT>
void Configurator<nodeT>::DeclareParameters(
    const std::vector<std::string> & parameter_full_names,
    const std::vector<rclcpp::ParameterType> & parameter_types
)
{
    declareParameters(parameter_full_names, parameter_types);

    std::unique_lock<std::shared_mutex> lock(parameters_mutex_);
    for (const auto & full_name : parameter_full_names) {
        if (std::find(managed_parameter_names_.begin(), managed_parameter_names_.end(), full_name) == managed_parameter_names_.end()) {
            managed_parameter_names_.push_back(full_name);
        }
    }
}

template <typename nodeT>
Configuration::SharedPtr Configurator<nodeT>::CreateConfiguration(
    const std::string & name,
    const std::vector<configuration_entry_t> & entries
)
{
    auto configuration = std::make_shared<Configuration>(
        name,
        entries,
        [this](const std::string & full_name) {
            rclcpp::Parameter parameter;
            if (!node_->get_parameter(full_name, parameter)) {
                throw std::runtime_error("Configuration getter could not read parameter " + full_name);
            }
            return parameter;
        }
    );

    std::unique_lock<std::shared_mutex> lock(parameters_mutex_);
    configurations_.push_back(configuration);
    return configuration;
}

template <typename nodeT>
void Configurator<nodeT>::SyncParameters(const std::vector<std::string> & parameter_full_names)
{
    if (parameter_full_names.empty()) {
        validate();
        return;
    }

    auto candidate_values = getCurrentValuesWithSchemaDefaults();
    for (const auto & full_name : parameter_full_names) {
        schema_validator_.ValidateParameterValue(full_name, candidate_values.at(full_name), candidate_values, true);
    }
}

template <typename nodeT>
std::string Configurator<nodeT>::GetParameterTypeString(rclcpp::ParameterType parameter_type)
{
    return SchemaValidator::ParameterTypeToString(parameter_type);
}

template <typename nodeT>
rclcpp::ParameterType Configurator<nodeT>::GetParameterTypeFromString(const std::string & parameter_type)
{
    return SchemaValidator::ParameterTypeFromString(parameter_type);
}

template <typename nodeT>
void Configurator<nodeT>::PrintParameters() const
{
    for (const auto & parameter_name : managed_parameter_names_) {
        rclcpp::Parameter parameter;
        if (node_->get_parameter(parameter_name, parameter)) {
            RCLCPP_INFO(node_->get_logger(), "%s: %s", parameter_name.c_str(), parameter.value_to_string().c_str());
        }
    }
}

template <typename nodeT>
void Configurator<nodeT>::PrintConfigurations() const
{
    for (const auto & configuration : configurations_) {
        RCLCPP_INFO(node_->get_logger(), "%s", configuration->name().c_str());
    }
}

template <typename nodeT>
void Configurator<nodeT>::validate() const
{
    if (managed_parameter_names_.empty()) {
        return;
    }
    schema_validator_.ValidateParameterMap(getCurrentValuesWithSchemaDefaults(), true);

    if (!on_set_parameters_callback_handle_) {
        auto * self = const_cast<Configurator<nodeT> *>(this);
        self->on_set_parameters_callback_handle_ = node_->add_on_set_parameters_callback(
            std::bind(&Configurator<nodeT>::onSetParametersCallback, self, std::placeholders::_1)
        );
    }

    if (!managed_node_announced_) {
        const_cast<Configurator<nodeT> *>(this)->announceManagedNode();
    }
}

template <typename nodeT>
void Configurator<nodeT>::declareParameters(
    const std::vector<std::string> & parameter_full_names,
    const std::vector<rclcpp::ParameterType> & parameter_types
)
{
    if (parameter_full_names.size() != parameter_types.size()) {
        throw std::runtime_error("Configurator::declareParameters(): Names/types size mismatch");
    }

    for (std::size_t i = 0; i < parameter_full_names.size(); ++i) {
        const auto & full_name = parameter_full_names[i];
        const auto & schema_entry = schema_validator_.GetParameter(full_name);

        if (schema_entry.parameter_type != parameter_types[i]) {
            throw std::runtime_error("Configurator::declareParameters(): Type mismatch for " + full_name);
        }

        rcl_interfaces::msg::ParameterDescriptor descriptor;
        descriptor.name = full_name;
        descriptor.description = "Managed by iii_drone_configuration schema";
        descriptor.read_only = false;
        descriptor.dynamic_typing = false;

        if (!node_->has_parameter(full_name)) {
            node_->declare_parameter(full_name, schema_entry.default_value, descriptor, false);
        }
    }
}

template <typename nodeT>
bool Configurator<nodeT>::undeclareParameters()
{
    std::unique_lock<std::shared_mutex> lock(parameters_mutex_);
    managed_parameter_names_.clear();
    return true;
}

template <typename nodeT>
void Configurator<nodeT>::parameterEventCallback(rcl_interfaces::msg::ParameterEvent parameter_event)
{
    if (parameter_event.node != node_->get_fully_qualified_name()) {
        return;
    }

    if (after_parameter_change_callback_ == nullptr) {
        return;
    }

    for (const auto & parameter : parameter_event.changed_parameters) {
        if (std::find(managed_parameter_names_.begin(), managed_parameter_names_.end(), parameter.name) != managed_parameter_names_.end()) {
            after_parameter_change_callback_(rclcpp::Parameter::from_parameter_msg(parameter));
        }
    }
}

template <typename nodeT>
rcl_interfaces::msg::SetParametersResult Configurator<nodeT>::onSetParametersCallback(
    const std::vector<rclcpp::Parameter> & parameters
)
{
    rcl_interfaces::msg::SetParametersResult result;
    result.successful = true;

    auto candidate_values = getCurrentValuesWithSchemaDefaults();

    for (const auto & parameter : parameters) {
        if (std::find(managed_parameter_names_.begin(), managed_parameter_names_.end(), parameter.get_name()) == managed_parameter_names_.end()) {
            continue;
        }

        candidate_values[parameter.get_name()] = parameter.get_parameter_value();

        try {
            schema_validator_.ValidateParameterValue(
                parameter.get_name(),
                parameter.get_parameter_value(),
                candidate_values,
                false
            );
        } catch (const std::exception & ex) {
            result.successful = false;
            result.reason = ex.what();
            return result;
        }
    }

    try {
        schema_validator_.ValidateParameterMap(candidate_values, true);
    } catch (const std::exception & ex) {
        result.successful = false;
        result.reason = ex.what();
        return result;
    }

    return result;
}

template <typename nodeT>
std::unordered_map<std::string, rclcpp::ParameterValue> Configurator<nodeT>::getCurrentManagedParameterValues() const
{
    std::unordered_map<std::string, rclcpp::ParameterValue> values;
    for (const auto & parameter_name : managed_parameter_names_) {
        rclcpp::Parameter parameter;
        if (!node_->get_parameter(parameter_name, parameter)) {
            throw std::runtime_error("Configurator::getCurrentManagedParameterValues(): Missing parameter " + parameter_name);
        }
        values.emplace(parameter_name, parameter.get_parameter_value());
    }
    return values;
}

template <typename nodeT>
std::unordered_map<std::string, rclcpp::ParameterValue> Configurator<nodeT>::getCurrentValuesWithSchemaDefaults() const
{
    std::unordered_map<std::string, rclcpp::ParameterValue> values;
    for (const auto & [parameter_name, schema_entry] : schema_validator_.parameters()) {
        values.emplace(parameter_name, schema_entry.default_value);
    }

    for (const auto & [parameter_name, value] : getCurrentManagedParameterValues()) {
        values[parameter_name] = value;
    }

    return values;
}

template <typename nodeT>
std::string Configurator<nodeT>::resolveSchemaFilePath()
{
    if (const char * explicit_file = std::getenv("III_DRONE_SCHEMA_FILE"); explicit_file != nullptr && explicit_file[0] != '\0') {
        return ExpandHome(explicit_file);
    }

    const auto package_share = ament_index_cpp::get_package_share_directory(
        "iii_drone_configuration"
    );
    const auto installed_schema = std::filesystem::path(package_share) /
        "configuration_contract" / "schema" / "parameter_manifest.yaml";
    if (!std::filesystem::is_regular_file(installed_schema)) {
        throw std::runtime_error(
            "Installed immutable configuration schema is unavailable: " +
            installed_schema.string()
        );
    }
    return installed_schema.string();
}

template <typename nodeT>
void Configurator<nodeT>::announceManagedNode() const
{
    if (!managed_node_announcement_publisher_) {
        return;
    }

    std_msgs::msg::String message;
    message.data = node_->get_fully_qualified_name();
    managed_node_announcement_publisher_->publish(message);
    const_cast<Configurator<nodeT> *>(this)->managed_node_announced_ = true;
}

template class iii_drone::configuration::Configurator<rclcpp::Node>;
template class iii_drone::configuration::Configurator<rclcpp_lifecycle::LifecycleNode>;
